src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,258 @@
+module Streamly.Coreutils.Cp
+    ( CpOptions (..)
+    , defaultCpOptions
+    , copy
+    , traverseDir
+    , cpFile
+    , cpFiles
+    , cpDirSerial
+    , cpDirAsync
+    , cpDirParallel
+   )
+where
+
+import Path
+import Streamly
+
+import qualified Streamly.Prelude as S
+import qualified Streamly.Internal.Prelude as IP
+import qualified Streamly.Internal.FileSystem.File as File
+
+import System.IO.Unsafe (unsafePerformIO)
+import System.Directory (listDirectory, doesFileExist, createDirectoryIfMissing)
+
+-------------------------------------------------------------------------------
+-- Record for options used with cp
+-------------------------------------------------------------------------------
+
+-- | Data type to represent flags\/options in GNU @cp@
+--
+-- @since 0.1.0.0
+data CpOptions = CpOptions
+    { verbose :: Bool
+    -- ^ Prints the source and destination files being copied
+    , force :: Bool
+    -- ^ if the destination file
+    -- cannot be opened, removes it and continues copying the rest
+    , overwriteExisting :: Bool
+    -- ^ If @True@, an existing destination file will be overwritten
+    -- due to copying
+    }
+
+
+-- | Default options for @cp@ are @verbose@ and @overwriteExisting@ to
+-- @True@ and @force@ set to @False@
+--
+-- @since 0.1.0.0
+{-# INLINE defaultCpOptions #-}
+defaultCpOptions :: CpOptions
+defaultCpOptions = CpOptions True False True
+
+
+-------------------------------------------------------------------------------
+-- cp and helper functions for options
+-------------------------------------------------------------------------------
+
+
+-- | Prints paths of source, destination files when @verbose@ is @True@.
+-- Returns @True@ if copying operation was successful.
+-- @False@ is returned in cases where a file by the same name
+-- exists in the destination path and @overwriteExisting@ is set to
+-- @False@ or if the source file does not exist.
+--
+-- @since 0.1.0.0
+{-# INLINE copy #-}
+copy
```

Comment:
This function`copy` is being used to map and copy files in a stream, if this function results in an IOError either due to not having the necessary permissions or failure to create a file the entire program crashes. But I think the correct thing to do here is to notify the error message to the user and continue copying rest of the files.

Reply:
depends on whether the `force` flag is true or not. If it is, then we should continue, else not.

Reply:
```bash
[pranaysashank issue]$ ls -l
total 0
---------- 1 pranaysashank pranaysashank 0 Aug 17 12:39 A
-rw-r--r-- 1 pranaysashank pranaysashank 0 Aug 17 12:39 B
-rw-r--r-- 1 pranaysashank pranaysashank 0 Aug 17 12:39 C
[pranaysashank issue]$ mkdir dest
[pranaysashank issue]$ cp B A C dest/
cp: cannot open 'A' for reading: Permission denied
[pranaysashank issue]$ ls dest/
B  C
```
Look at this example on my machine, it failed to copy A because it doesn't have the permission to read, but it continued copying the remaining files i.e. C